### ISSUE TITLE
[Multi-device] ```jnp.mean()``` on sharded tensor followed by all-reduce with ```lax.pmean()``` produces incorrect results

### ISSUE THREAD

**Issue by umalesTT** [tenstorrent/tt-xla#613]
### Describe the bug

During data-parallel MNIST training in JAX, at the end of a training step loss is calculated by averaging local losses of all used devices with ```lax.pmean()```, where one local loss is calculated by ``` jnp.mean( (logits - labels) ** 2)```. This ends up resulting in loss being ```inf```.

Below are stablehlo and ttir graphs for a minimal example that reproduces this error.

```
module @jit__lambda_ attributes {mhlo.num_partitions = 2 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<4x4xf32> {mhlo.sharding = "{devices=[2,1]<=[2]}"}) -> (tensor<f32> {jax.result_info = "", mhlo.sharding = "{replicated}"}) {
    %0 = stablehlo.custom_call @Sharding(%arg0) {mhlo.sharding = "{devices=[2,1]<=[2]}"} : (tensor<4x4xf32>) -> tensor<4x4xf32>
    %1 = stablehlo.custom_call @SPMDFullToShardShape(%0) {mhlo.sharding = "{manual}"} : (tensor<4x4xf32>) -> tensor<2x4xf32>
    %2 = call @shmap_body(%1) : (tensor<2x4xf32>) -> tensor<f32>
    %3 = stablehlo.custom_call @Sharding(%2) {mhlo.sharding = "{manual}"} : (tensor<f32>) -> tensor<f32>
    %4 = stablehlo.custom_call @SPMDShardToFullShape(%3) {mhlo.sharding = "{replicated}"} : (tensor<f32>) -> tensor<f32>
    return %4 : tensor<f32>
  }
  func.func private @shmap_body(%arg0: tensor<2x4xf32>) -> (tensor<f32> {jax.result_info = "[]"}) {
    %cst = stablehlo.constant dense<2.000000e+00> : tensor<f32>
    %cst_0 = stablehlo.constant dense<8.000000e+00> : tensor<f32>
    %cst_1 = stablehlo.constant dense<0.000000e+00> : tensor<f32>
    %0 = stablehlo.reduce(%arg0 init: %cst_1) applies stablehlo.add across dimensions = [0, 1] : (tensor<2x4xf32>, tensor<f32>) -> tensor<f32>
    %1 = stablehlo.divide %0, %cst_0 : tensor<f32>
    %2 = "stablehlo.all_reduce"(%1) <{channel_handle = #stablehlo.channel_handle<handle = 1, type = 0>, replica_groups = dense<[[0, 1]]> : tensor<1x2xi64>, use_global_device_ids}> ({
    ^bb0(%arg1: tensor<f32>, %arg2: tensor<f32>):
      %4 = stablehlo.add %arg1, %arg2 : tensor<f32>
      stablehlo.return %4 : tensor<f32>
    }) : (tensor<f32>) -> tensor<f32>
    %3 = stablehlo.divide %2, %cst : tensor<f32>
    return %3 : tensor<f32>
  }
}
```

```
module @jit__lambda_ attributes {mhlo.num_partitions = 2 : i32, mhlo.num_replicas = 1 : i32, tt.meshes = #tt.meshes<[<"mesh_gspmd" = 1x2>]>} {
  func.func public @main(%arg0: tensor<4x4xf32>) -> (tensor<1xf32> {jax.result_info = ""}) {
    %0 = "ttir.constant"() <{value = dense<0.000000e+00> : tensor<1xf32>}> : () -> tensor<1xf32>
    %1 = "ttir.constant"() <{value = dense<8.000000e+00> : tensor<1xf32>}> : () -> tensor<1xf32>
    %2 = "ttir.constant"() <{value = dense<2.000000e+00> : tensor<1xf32>}> : () -> tensor<1xf32>
    %3 = ttir.empty() : tensor<2x4xf32>
    %4 = "ttir.mesh_shard"(%arg0, %3) <{shard_dims = array<i64: -1, 0>, shard_direction = #tt.shard_direction<full_to_shard>, shard_shape = array<i64: 2, 1>, shard_type = #tt.shard_type<identity>}> : (tensor<4x4xf32>, tensor<2x4xf32>) -> tensor<2x4xf32>
    %5 = ttir.empty() : tensor<1xf32>
    %6 = "ttir.sum"(%4, %5) <{dim_arg = [0 : i32, 1 : i32], keep_dim = false}> : (tensor<2x4xf32>, tensor<1xf32>) -> tensor<1xf32>
    %7 = ttir.empty() : tensor<1xf32>
    %8 = "ttir.div"(%6, %1, %7) : (tensor<1xf32>, tensor<1xf32>, tensor<1xf32>) -> tensor<1xf32>
    %9 = ttir.empty() : tensor<1xf32>
    %10 = "ttir.all_reduce"(%8, %9) <{cluster_axis = 1 : ui32, reduce_type = #tt.reduce_type<sum>}> : (tensor<1xf32>, tensor<1xf32>) -> tensor<1xf32>
    %11 = ttir.empty() : tensor<1xf32>
    %12 = "ttir.div"(%10, %2, %11) : (tensor<1xf32>, tensor<1xf32>, tensor<1xf32>) -> tensor<1xf32>
    return %12 : tensor<1xf32>
  }
}
```

### Steps to reproduce the issue

Run 
```
import jax
import jax.numpy as jnp
import numpy as np
from jax.sharding import Mesh, PartitionSpec, NamedSharding
from jax.experimental import shard_map
import os
import sys
import jax._src.xla_bridge as xb

def init_device():
    priority = 500
    backend = "tt"
    path = os.path.join(os.getcwd(), "build/src/tt/pjrt_plugin_tt.so")
    if not os.path.exists(path):
        raise FileNotFoundError(f"Could not find tt_pjrt C API plugin at {path}")

    print("Loading tt_pjrt C API plugin", file=sys.stderr)
    xb.discover_pjrt_plugins()

    plugin = xb.register_plugin(backend, priority=priority, library_path=path, options=None)
    print("Loaded", file=sys.stderr)
    jax.config.update("jax_platforms", "cpu,tt")

def main():
    init_device()
    #jax.config.update("jax_use_shardy_partitioner", True)
    devices = np.array(jax.devices("tt"))
    mesh = Mesh(devices, axis_names=("dp",))

    data = jnp.arange(16, dtype=jnp.float32).reshape(4, 4)

    # We shard the array across 2 devices along the first dimension
    sharding = NamedSharding(mesh, PartitionSpec("dp", None))
    sharded_data = jax.device_put(data, sharding)

    try:
        sharded_mean = shard_map.shard_map(
            lambda x: jax.lax.pmean(jnp.mean(x), axis_name="dp"),
            mesh=mesh,
            in_specs=PartitionSpec("dp", None),
            out_specs=PartitionSpec(),
            check_rep=False,
        )
        sharded_mean_jit = jax.jit(sharded_mean, out_shardings=NamedSharding(mesh, PartitionSpec()))
        result = sharded_mean_jit(sharded_data)
        print(f"Mean of sharded data: {result}")
        #print(type(result))
    except Exception as e:
        print(f"Error computing mean on sharded data: {e}")

if __name__ == "__main__":
    main()
```

**Comment by umalesTT** [comment]
In the provided ttir code we may see that the output has no information regarding sharding.
